base/applications/rapps/asyncinet.cpp

```diff
@@ -0,0 +1,258 @@
+                        break;
+                    }
+                    else
+                    {
+                        //something went wrong
+                        if (dwError == ERROR_INVALID_HANDLE)
+                        {
+                            // That's most likely not an error. handle got closed when canceling may lead to this
+                            break;
+                        }
+                        if (AsyncInet->Callback)
+                        {
+                            AsyncInet->Callback(AsyncInet, ASYNCINET_ERROR, 0, (LPARAM)dwError, AsyncInet->Extension);
+                        }
+                        break;
+                    }
+                }
+            }
+        }
+        break;
+        case ERROR_INVALID_HANDLE:
+            if (AsyncInet->bIsCancelled)
+            {
+                // this is most likely not an error. when canceling, using closed/invalid handle may lead to this
+                break;
+            }
+            // fall-down to handle error
+        default:
+            // something went wrong
+            AsyncInetPerformCallback(AsyncInet, ASYNCINET_ERROR, 0, (LPARAM)(AsyncResult->dwError));
+            break;
+        }
+        break;
+    }
+
+    }
+    return;
+}
+
+
+pASYNCINET AsyncInetDownload(LPCWSTR lpszAgent,
+    DWORD   dwAccessType,
+    LPCWSTR lpszProxy,
+    LPCWSTR lpszProxyBypass,
+    LPCWSTR lpszUrl,
+    BOOL bAllowCache,
+    ASYNCINET_CALLBACK Callback,
+    VOID* Extension
+    )
+{
+    pASYNCINET AsyncInet = NULL;
+    BOOL bSuccess = FALSE;
+    DWORD InetOpenUrlFlag = 0;
+    INTERNET_STATUS_CALLBACK OldCallbackFunc;
+
+    AsyncInet = (pASYNCINET)HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(ASYNCINET));
+    if (!AsyncInet)goto cleanup;
+
+    AsyncInet->Callback = Callback;
+    AsyncInet->Extension = Extension;
+
+    AsyncInet->hInternet = InternetOpenW(lpszAgent, dwAccessType, lpszProxy, lpszProxyBypass, INTERNET_FLAG_ASYNC);
+    if (!AsyncInet->hInternet)goto cleanup;
+
+    OldCallbackFunc = InternetSetStatusCallbackW(AsyncInet->hInternet, AsyncInetStatusCallback);
+    if (OldCallbackFunc == INTERNET_INVALID_STATUS_CALLBACK)goto cleanup;
+
+    InetOpenUrlFlag |= INTERNET_FLAG_KEEP_CONNECTION | INTERNET_FLAG_PASSIVE;
+    if (!bAllowCache)
+    {
+        InetOpenUrlFlag |= INTERNET_FLAG_DONT_CACHE | INTERNET_FLAG_PRAGMA_NOCACHE;
+    }
+    AsyncInet->hInetFile = InternetOpenUrlW(AsyncInet->hInternet, lpszUrl, 0, 0, InetOpenUrlFlag, (DWORD_PTR)AsyncInet);
+    if (!AsyncInet->hInetFile)
+    {
+        if (GetLastError() != ERROR_IO_PENDING)
+        {
+            goto cleanup;
+        }
+    }
+    else
+    {
+        // TODO: If I remember it correctly, sometimes the file is cache before
+        // thus may lead to InternetOpenUrlW return immediately with the handle.
+        // more investigation required. And if it's true, it should be handled correctly.
```

Comment:
Most image hosts (like imgur) use image caching,
otherwise something like this might work: https://refreshyourcache.com/en/cache-test/

Reply:
It's not about caching, it about will InternetOpenUrl still has a chance to directly return with the handle (instead of return 0 with GetLastError() == IO_PENDING) even if I specify INTERNET_FLAG_ASYNC in InternetOpen (the entry of InternetOpenUrl in msdn didn't elaborate this, but I do encountered this situation years ago). I just need to seek for confirmation

Reply:
yeah this should be taken into account.

Reply:
Great, I have found the evidence.
https://docs.microsoft.com/zh-cn/windows/win32/wininet/common-functions#setting-up-asynchronous-operations
> Therefore, depending on the request, the operation might complete synchronously or asynchronously. The application should check the return code. If a function returns FALSE or NULL, and GetLastError returns ERROR_IO_PENDING, the request has been made asynchronously

I'm going to improve this now